traitsui/testing/tester/ui_tester.py

```diff
@@ -0,0 +1,260 @@
+#  Copyright (c) 2005-2020, Enthought, Inc.
+#  All rights reserved.
+#
+#  This software is provided without warranty under the terms of the BSD
+#  license included in LICENSE.txt and may be redistributed only
+#  under the conditions described in the aforementioned license.  The license
+#  is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+#  Thanks for using Enthought open source!
+#
+
+from traitsui.ui import UI
+from traitsui.testing.tester import locator
+from traitsui.testing.tester.registry import LocationRegistry
+from traitsui.testing.tester.ui_wrapper import UIWrapper
+from traitsui.tests._tools import (
+    create_ui as _create_ui,
+)
+
+
+class UITester:
+    """ UITester assists testing of GUI applications developed using TraitsUI.
+
+    The following actions typically found in tests are supported by the tester:
+
+    - (1) Create a GUI that will be cleaned up when the test exits.
+    - (2) Locate the GUI element to be tested.
+    - (3) Perform the user interaction for side effects (e.g. mouse clicking)
+    - (4) Inspect GUI element as a user would (e.g. checking the displayed text
+          on a widget)
+
+    Creating a GUI
+    --------------
+    Given a ``HasTraits`` object, a GUI can be created using the
+    ``UITester.create_ui`` method::
+
+        class App(HasTraits):
+            text = Str()
+
+        obj = App()
+        tester = UITester()
+        with tester.create_ui(obj) as ui:
+            pass
+
+    ``create_ui`` is a context manager such that it ensures the GUI is always
+    disposed of at the end of a test.
+
+    The returned value is an instance of ``traitsui.ui.UI``. This is the entry
+    point for locating GUI elements for further testing.
+
+    Locating GUI elements
+    ---------------------
+    After creating an ``UI`` object, ``UITester.find_by_name`` can be used
+    to locate a specific UI target::
+
+        obj = App()
+        tester = UITester()
+        with tester.create_ui(obj) as ui:
+            wrapper = tester.find_by_name(ui, "text")
+
+    The returned value is an instance of ``UIWrapper``. It wraps the
+    UI target instance found and allows further test actions to be performed on
+    the target.
+
+    Since it is fairly typical for a UI target to have a nested UI (and those
+    nested UI may bave more nested UI), to locate these nested UIs,
+    ``UIWrapper.find_by_name`` can be used::
+
+        class Person(HasTraits):
+            name = Str()
+
+        class Account(HasTraits):
+            person = Instance(Person)
+
+        view = View(Item(name="person", style="custom"))
+        account = Account(person=Person())
+        tester = UITester()
+        with tester.create_ui(account, dict(view=view)) as ui:
+            wrapper = tester.find_by_name(ui, "person").find_by_name("name")
```

Comment:
This will raise on this branch because I moved the logic to the the LocationRegistry and as an effort to reduce the scope, I removed the solver for getting the nested UI in a custom InstanceEditor. I should remove this description.

The implementation can be added back later once we have the toolkit specific default registry set up.
See for example:
https://github.com/enthought/traitsui/blob/ceb8fee665c741b92dc9ff8e5e9c36ef83f206a8/traitsui/testing/qt4/default_registry.py#L109-L113
https://github.com/enthought/traitsui/blob/ceb8fee665c741b92dc9ff8e5e9c36ef83f206a8/traitsui/testing/qt4/default_registry.py#L48-L49